### PR TITLE
[Paddle-TRT] constant-folding

### PR DESCRIPTION
**Pull request by zhoutianzi666**
### PR types

Others
### PR changes

Others
### Describe

 When a op's inputs and outputs is determined before feeding data to the model, we can remove this op from the model. 
This ConstantFolding pass can remove all these like ops.

### CODE REVIEW

File: paddle/fluid/inference/tensorrt/convert/conv2d_op.cc [hunk 1]
```diff
@@ -138,12 +142,13 @@ void ConvertConv2d(TensorRTEngine* engine,
       platform::errors::Fatal("TensorRT create conv2d/conv2d_transpose"
                               " layer failed."));
   layer->setStride(nv_strides);
-  if (paddings.size() == 2) {
-    layer->setPadding(nv_paddings);
-  } else {
-    layer->setPrePadding(nv_pre_paddings);
-    layer->setPostPadding(nv_post_paddings);
+
+  layer->setPrePadding(nv_pre_paddings);
```
Comment (zhangjun) [inline paddle/fluid/inference/tensorrt/convert/conv2d_op.cc]: 接口很快要废弃，试试使用[setPrePaddingNd](https://docs.nvidia.com/deeplearning/tensorrt/api/c_api/classnvinfer1_1_1_i_padding_layer.html#ada84a2215dbb0f8f1b040366b7400197)

File: python/paddle/fluid/tests/unittests/ir/inference/test_trt_convert_conv2d_transpose.py [hunk 2]
```diff
@@ -311,19 +311,19 @@ def generate_trt_nodes_num(attrs, dynamic_shape):
...
         # for dynamic_shape
         generate_dynamic_shape(attrs)
         self.trt_param.precision = paddle_infer.PrecisionType.Float32
         yield self.create_inference_config(), generate_trt_nodes_num(
-            attrs, True), 1e-5
+            attrs, True), 1e-4
         self.trt_param.precision = paddle_infer.PrecisionType.Half
         yield self.create_inference_config(), generate_trt_nodes_num(
-            attrs, True), (1e-5, 1e-3)
+            attrs, True), (1e0, 1e-3)
```
Comment (zhangjun) [inline python/paddle/fluid/tests/unittests/ir/inference/test_trt_convert_conv2d_transpose.py]: 精度有问题吗

File: paddle/fluid/framework/ir/constant_folding_pass.cc [hunk 3]
```diff
@@ -0,0 +1,234 @@
+/* Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserved.
```
Comment (qingqing01) [inline paddle/fluid/framework/ir/constant_folding_pass.cc]: 2022

File: paddle/fluid/framework/ir/constant_folding_pass.cc [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+namespace ir {
+class Node;
+}  // namespace ir
+}  // namespace framework
+}  // namespace paddle
+
+namespace paddle {
+namespace framework {
+namespace ir {
+namespace patterns {
```
Comment (qingqing01) [inline paddle/fluid/framework/ir/constant_folding_pass.cc:46]: Add more comments for new Pass in the beginning, such as patter relu

File: paddle/fluid/framework/naive_executor.cc [hunk 5]
```diff
@@ -53,6 +51,28 @@ void NaiveExecutor::Run() {
...
+      if (var->persistable) {
+      //  std::cout << name << std::endl;
+      }
+    }
+    for (size_t i = 0; i < op->OutputVars(true).size() * 0; i++) {
+      auto name = op->OutputVars(true)[i];
+      auto *var = scope_->FindVar(name);
+      var->persistable = all_input_persistable;
+    }
+    //std::cout << "end" << std::endl;
```
Comment (qingqing01) [inline paddle/fluid/framework/naive_executor.cc]: 这些是debug code？

File: paddle/fluid/framework/variable.h [hunk 6]
```diff
@@ -73,6 +73,8 @@ class Variable {
     return holder_->Type();
   }
 
+  bool persistable = false;
```
Comment (qingqing01) [inline paddle/fluid/framework/variable.h]: Why need this attribute？

File: paddle/fluid/inference/tensorrt/op_teller.cc [hunk 7]
```diff
@@ -1986,10 +1986,11 @@ bool OpTeller::Tell(const framework::ir::Node* node,
     }
 #endif
 
-    // conv2d_transpose, conv3d_transpose, depthwise_conv2d_transpose
-    if (op_type.find("d_transpose") > 0) {
-      // trt doen't support output_padding,
-      // output_padding is set when stride > 1
+    // conv3d_transpose
+    if (op_type.find("3d_transpose") != op_type.npos) {
```
Comment (qingqing01) [inline paddle/fluid/inference/tensorrt/op_teller.cc]: Look at the original code,  support 2d_transpose、3d_transpose..., now only find 3d_tranpose？

File: python/paddle/fluid/tests/unittests/ir/inference/test_trt_convert_conv3d_transpose.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+        pass
+
+    def test(self):
+        self.add_skip_trt_case()
+        self.run_test()
+
+    def test_quant(self):
+        self.add_skip_trt_case()
+        self.run_test(quant=True)
+
```
Comment (qingqing01) [inline python/paddle/fluid/tests/unittests/ir/inference/test_trt_convert_conv3d_transpose.py]: need testing all supported constnt-folding case

**Comment by zhoutianzi666** [comment]
我关掉它

**Pull request closed without merging**